plugins/sslservices/sslservices.cpp

```diff
@@ -0,0 +1,934 @@
+        mdCtx = EVP_MD_CTX_new();
+        if (!mdCtx)
+            failOpenSSLError("EVP_MD_CTX_new (pkSign)");
+
+        const EVP_MD *md = digestCache.checkCache(_algorithm_name);
+
+        if (EVP_DigestSignInit(mdCtx, nullptr, md, nullptr, privateKey) <= 0)
+            failOpenSSLError("EVP_DigestSignInit (pkSign)");
+
+        // Add plaintext to context
+        if (EVP_DigestSignUpdate(mdCtx, _plaintext, len_plaintext) <= 0)
+            failOpenSSLError("EVP_DigestSignUpdate (pkSign)");
+
+        // Determine the buffer length for the signature
+        size_t signatureLen = 0;
+        if (EVP_DigestSignFinal(mdCtx, nullptr, &signatureLen) <= 0)
+            failOpenSSLError("determining result length (pkSign)");
+
+        // Allocate memory for the signature
+        MemoryBuffer signatureBuffer(signatureLen);
+
+        // Perform the actual signing
+        if (EVP_DigestSignFinal(mdCtx, static_cast<byte *>(signatureBuffer.bufferBase()), &signatureLen) <= 0)
+            failOpenSSLError("signing (pkSign)");
+
+        // Set the result
+        __lenResult = signatureLen;
+        __result = signatureBuffer.detachOwn();
+
+        // Clean up
+        EVP_MD_CTX_free(mdCtx);
+    }
+    catch (...)
+    {
+        if (mdCtx)
+            EVP_MD_CTX_free(mdCtx);
+        rtlFree(__result);
+        __lenResult = 0;
+        throw;
+    }
+}
+
+SSLSERVICES_API bool SSLSERVICES_CALL pkVerifySignature(ICodeContext *ctx, size32_t len_signature, const void * _signature, size32_t len_signedData, const void * _signedData, size32_t len_pem_public_key, const char * _pem_public_key, const char * _algorithm_name)
+{
+    EVP_MD_CTX *mdCtx = nullptr;
+
+    try
+    {
+        // Load the public key from the PEM string
+        EVP_PKEY * publicKey = pkeyCache.checkCache(len_pem_public_key, _pem_public_key, 0, nullptr);
+
+        // Create and initialize the message digest context
+        mdCtx = EVP_MD_CTX_new();
+        if (!mdCtx)
+            failOpenSSLError("EVP_MD_CTX_new");
+
+        const EVP_MD *md = digestCache.checkCache(_algorithm_name);
+
+        if (EVP_DigestVerifyInit(mdCtx, nullptr, md, nullptr, publicKey) <= 0)
+            failOpenSSLError("EVP_DigestVerifyInit (pkVerifySignature)");
+
+        if (EVP_DigestVerifyUpdate(mdCtx, _signedData, len_signedData) <= 0)
+            failOpenSSLError("EVP_DigestVerifyUpdate (pkVerifySignature)");
+
+        // Perform the actual verification
+        size32_t res = EVP_DigestVerifyFinal(mdCtx, reinterpret_cast<const unsigned char *>(_signature), len_signature);
+
+        // Clean up
+        EVP_MD_CTX_free(mdCtx);
+        return res == 1;
+    }
+    catch (...)
+    {
+        if (mdCtx)
+            EVP_MD_CTX_free(mdCtx);
+        throw;
+    }
+}
+
+MODULE_INIT(INIT_PRIORITY_STANDARD)
+{
+    pkeyCache.init();
+    digestCache.init();
+    cipherCache.init();
+
+    return true;
+}
+
+MODULE_EXIT()
+{
+    if(PRINT_STATS)
+    {
+        pkeyCache.printStatistics();
```

Comment:
This is only going to work for hthor - because the caches are threaded local.  Even then there are situations where it will display nothing.